Documentation changes

DHager opened the pull request:
Took a try at editing the README because I'm not quite sure how else to contribute. 

Certain spots need fact-checking to make sure I understood correctly:
- Does the object hash change only after the _first_ `Weak\Reference`, or does it change every time a _new_ `Weak\Reference` is created? Are there any other circumstances where it may change further?
- I changed the php-weak-stubs composer settings to match the names I saw on packagist, but haven't tested it directly.

I've taken out some technical detail in order to shorten the the README, since most readers will probably be writers of code in PHP rather than extension-writers.
- Removed the part about the pre-release implementation storing references in an array, because it's less-likely that the README readers will need to know those tradeoffs -- they can be told in a github-issue discussion.
- Removed some of the technical object-hash/object-handler-hash details about how the `spl_object_hash()`, or parts of it,  can change.
- Removed the link to the [php-weak-stubs](https://github.com/pinepain/php-weak-stubs) github repository. Anyone reporting issues or making PRs should be here at the main site, right?

pinepain committed:
Initial commit

DHager committed:
Update README.md

Rewording/editing because I'm not quite sure how else to contribute. 

I've also taken out some technical detail in order to shorten the the README, since most users are probably deal with PHP code rather than the internals:

* I removed the part about the pre-release implementation storing references in an array, because it's less-likely that the README readers will need to know those tradeoffs -- they can be told in a github-issue discussion.
* Removed some of the details about how the `spl_object_hash()` can change
* Removed the link to the [php-weak-stubs](https://github.com/pinepain/php-weak-stubs) github repository, since most people should be reporting issues on the main project instead.

Certain spots need fact-checking to make sure I understood correctly:

* Does the object hash change only after the *first* `Weak\Reference`, or does it change every time a *new* `Weak\Reference` is created? Are there any other circumstances where it may change further?
* I changed the php-weak-stubs composer settings to match what I saw on packagist, but haven't tested it yet.

DHager committed:
Update README.md

Adding a bit to basic usage, putting some stub-PHP links on the class/function list since that's where some people will want to click.

pinepain commented on README.md:
````diff
@@ -79,16 +81,15 @@ $ref2 = new Reference($obj, function () {});
 ```
 
-Strict comparison works as always:
+Strict comparison should still work as expected:
````
Both cases are expected =). I originally tried to point that I overloaded compare object handler, so now loose comparison counts hidden from props referent object, but not notifier. Strict comparison is a core php feature and afaik it compare object ce and couldn't be overloaded.

pinepain commented on README.md:
```diff
@@ -175,26 +171,22 @@ foreach($weak_references as $weak_ref_object_handle => $weak_reference) {
-For plumbing memory-related problems there are valgrind, to activate it, execute `export TEST_PHP_ARGS=-m` before running tests.
+For debugging memory-related problems, valgrind is already installed. To activate it, execute `export TEST_PHP_ARGS=-m` before running tests.
 
-To prevent asking test suite to send results to PHP QA team, `NO_INTERACTION=1` env variable is set. If run tests in your
```
Here I tried to say that out of the box in .bashrc NO_INTERACTION env variable is already set to 1

pinepain commented on README.md:
```diff
@@ -175,26 +171,22 @@ foreach($weak_references as $weak_ref_object_handle => $weak_reference) {
-of date sometimes) for type-hinting inside your IDE in case you don't have PHP 7 include files installed. They are
-provided only for typehinting and nothing more. While for quick fixes they are ok, for real development you have to point
-to real file from the same PHP version you are building extension for.
+There is a `php7debugzts` directory inside this repo which contains include files for the master php branch (which may be out of date sometimes) to support type-hinting inside your IDE in case you don't have PHP7 include files installed. They are provided only as a convenience, and any serious deveopment should use the real files from the exact PHP version you are developing against.
```
I have feedback from php internals devs that it is not very wise idea to ship it and I agree with them. It also pollute diffs significantly, so I guess I will remove it and force-push repo (incl tags), it is wrong, but as long as no functionality touched, i guess, it is ok.

pinepain commented on README.md:
```diff
@@ -1,70 +1,72 @@
-Anyway, we assume that exception in destructors are fatal and the only way they should be handled is fail-fast approach.
- 
-If you fall down under this limitation and have to try-catch object destructing, pleas, fill an issue and we'll discuss
-how this issue may be solved. One of pre-release features was storing destructed object's weak reference in some array
```
I would still love to mention it as if someone read it I can get response on this idea. Currently I have few negative responses, but I would love to hear more of them, so not sure whether this still be removed completely.

DHager replied:
I've opened issue #2 so that if someone is interested there's an immediate place they can comment and see previous comments.

pinepain commented on README.md:
```diff
@@ -1,70 +1,72 @@
-#### Object's `spl_object_hash()` hash changes after `Weak\Reference` created for it.
+Creating a `Weak\Reference` has a side-effect on on the object that you are making the reference to. Because it changes certain internal PHP pointers, it will alter the `spl_object_hash()` of the object. 
 
-Internally `Weak\Reference` changes object handlers pointer, so `spl_object_hash()` function will provide different hash
```
Agree that mention such details may be out of scope, but weak references feature targets pretty experienced developers. While it is probably out of readme scope at this time I prefer to keep it.

DHager replied:
Hopefully it'll be fixed in php-src and nobody will need to care :)

pinepain replied:
PHP 7.0.0, 7.0.1 and 7.0.2 already released, so there are some chance to surprise users. I will mention that in readme, smth like PHP 7 required, PHP > 7.0.2 required and then explanation why.

pinepain commented on README.md:
```diff
@@ -103,29 +104,23 @@ $ref2 = clone $ref1;
+If you are also using [Composer](https://getcomposer.org/), it is recommended that you add the `pinepain/php-weak-stubs` package as a dev-mode requirement. This provides skeleton definitions and annotations to enable support for auto-completion in your IDE and other code-analysis tools.
 
-    composer require --dev php-weak-stubs
+    composer require --dev pinepain/php-weak-stubs
```
Yes, this is correct package name, missed repo owner for some reason.

pinepain commented:
Thanks for your help!

I commented on diff, tried to explain some ideas.
Can you also look into stubs to see whether comments there are informative?

> Does the object hash change only after the first Weak\Reference, or does it change every time a new Weak\Reference is created?

It changes only once after the first `Weak\Reference` created.

> Are there any other circumstances where it may change further?

After we substitue object handlers table, we set it back only before object death, so i guess it is almost the same as not changes.

I have discussion with php internals devs and we comes to a solution to just remove object handlers pointer from `spl_object_hash()` generation to make hash consistent. Here is PR https://github.com/php/php-src/pull/1724 for that, but not ETA or even guarantees that it will be merged (but I really hope it will be), weakref extension also needs that.

DHager commented:
Personally, the mutation of `spl_object_hash()` is the limitation that worries me the most, since it can easily break entirely-unrelated code as soon as someone creates a weak-reference. So yeah, good to see that php/php-src#1724 just got merged, let's hope it finds its way into a PHP7 release soon.

DHager committed:
Update README.md

Extracting some data into issue #2
Removing notes about loose equality for reference-objects, since it's not really a surprise bug or limitation, will move it to the stub-documentation.

pinepain commented:
JFYI: when editing text files that doesn't affect functionality you can add `[skip ci]` to the commit message so that tests will be skipped, e.g. "Update README.md [skip ci]".

DHager committed:
Clarify NO_INTERACTION env-var setting

DHager committed:
Adding note about loose-equality

... after removing it from the README.

pinepain commented:
Automatically closed without ability to reopen due to force-pushing the whole repo to remove dev files. Will try to get back to this changes later.
